roofit/roofitcore/src/TestStatistics/RooAbsL.cxx

```diff
@@ -0,0 +1,456 @@
+
+   //   cout << "ROATS::ctor(" << GetName() << ") funcClone structure dump AFTER opt" << endl ;
+   //   pdf->Print("t") ;
+
+   // optimization steps (copied from ROATS::optimizeCaching)
+
+   pdf_->getVal(_normSet.get());
+   // Set value caching mode for all nodes that depend on any of the observables to ADirty
+   pdf_->optimizeCacheMode(*_funcObsSet);
+   // Disable propagation of dirty state flags for observables
+   data_->setDirtyProp(kFALSE);
+
+   // Disable reading of observables that are not used
+   data_->optimizeReadingWithCaching(*pdf_, RooArgSet(), RooArgSet()) ;
+
+}
+
+RooArgSet *RooAbsL::getParameters()
+{
+   auto ding = pdf_->getParameters(*data_);
+   return ding;
+
+//   // *** START HERE
+//   // WVE HACK determine if we have a RooRealSumPdf and then treat it like a binned likelihood
+//   RooAbsPdf *binnedPdf = 0;
+//   if (pdf_->getAttribute("BinnedLikelihood") && pdf_->IsA()->InheritsFrom(RooRealSumPdf::Class())) {
+//      // Simplest case: top-level of component is a RRSP
+//      binnedPdf = pdf_.get();
+//   } else if (pdf_->IsA()->InheritsFrom(RooProdPdf::Class())) {
+//      // Default case: top-level pdf is a product of RRSP and other pdfs
+//      RooFIter iter = ((RooProdPdf *)pdf_.get())->pdfList().fwdIterator();
+//      RooAbsArg *component;
+//      while ((component = iter.next())) {
+//         if (component->getAttribute("BinnedLikelihood") &&
+//             component->IsA()->InheritsFrom(RooRealSumPdf::Class())) {
+//            binnedPdf = (RooAbsPdf *)component;
+//         }
+//         if (component->getAttribute("MAIN_MEASUREMENT")) {
+//            // not really a binned pdf, but this prevents a (potentially) long list of subsidiary measurements to
+//            // be passed to the slave calculator
+//            binnedPdf = (RooAbsPdf *)component;
+//         }
+//      }
+//   }
+//   // WVE END HACK
+//
+//   std::unique_ptr<RooArgSet> actualParams {binnedPdf ? binnedPdf->getParameters(data_.get()) : pdf_->getParameters(data_.get())};
+//   RooArgSet* selTargetParams = (RooArgSet *)ding->selectCommon(*actualParams);
+//
+//   std::cout << "RooAbsL::getParameters:" << std::endl;
+//   selTargetParams->Print("v");
+//
+//   return selTargetParams;
+}
+
+void RooAbsL::constOptimizeTestStatistic(RooAbsArg::ConstOpCode opcode, bool doAlsoTrackingOpt)
+{
+   // to be further implemented, this is just a first test implementation
+   if (opcode == RooAbsArg::Activate) {
+      ConstantTermsOptimizer::enableConstantTermsOptimization(pdf_.get(), _normSet.get(), data_.get(), doAlsoTrackingOpt);
+   }
+}
+
+std::string RooAbsL::GetName() const
+{
+   std::string output("likelihood of pdf ");
+   output.append(pdf_->GetName());
+   return output;
+}
+
+std::string RooAbsL::GetTitle() const
+{
+   std::string output("likelihood of pdf ");
+   output.append(pdf_->GetTitle());
+   return output;
+}
+
+void RooAbsL::optimizePdf()
+{
+   // TODO: implement, using ConstantTermsOptimizer
```

Comment:
Try to add as few `TODO` as possible.

Reply:
Already did it, but called differently, removed `optimizePdf`.